src/xcode/ENA/ENA/Source/AppDelegate.swift

```diff
@@ -224,4 +226,76 @@ extension AppDelegate: ENATaskExecutionDelegate {
 			completion(true)
 		}
 	}
+
+	// MARK: - Plausible deniability.
+
+	var minHoursToNextBackgroundExecution: Double { 4 }
+	var maxHoursToNextBackgroundExecution: Double { 12 }
+	var numberOfDaysToRunPlaybook: Double { 16 }
+	var minNumberOfSequentialPlaybooks: Int { 1 }
+	var maxNumberOfSequentialPlaybooks: Int { 4 }
+	var minDelayBetweenSequentialPlaybooks: Int { 5 } // seconds
+	var maxDelayBetweenSequentialPlaybooks: Int { 10 } // seconds
+
+	/// Trigger a fake playbook to enable plausible deniability.
+	private func executeFakeRequests(_ completion: (() -> Void)? = nil) {
+		guard store.isAllowedToSubmitDiagnosisKeys else {
+			completion?()
+			return
+		}
+
+		// Initialize firstPlaybookExecution date during the first run regardless of actual execution.
+		if store.firstPlaybookExecution == nil {
+			store.firstPlaybookExecution = Date()
+		}
+
+		// Time interval until we want to resend a fake request from the background.
+		let offset = Double.random(in: minHoursToNextBackgroundExecution...maxHoursToNextBackgroundExecution) * 60
+		let now = Date()
+
+		if
+			let firstPlaybookExecution = store.firstPlaybookExecution,
+			firstPlaybookExecution.addingTimeInterval(numberOfDaysToRunPlaybook * 86_400) > now,
+			store.lastBackgroundFakeRequest.addingTimeInterval(offset) > now
+		{
+			sendFakeRequest {
+				self.store.lastBackgroundFakeRequest = now
+				completion?()
+			}
+		} else {
+			completion?()
+		}
+	}
+
+	func executeFakeRequestOnAppLaunch() {
+		// Execute a fake request 1 in 100 times while we are running in foreground.
+		guard
+			UIApplication.shared.applicationState == .active,
+			Int.random(in: 0...100) == 6
+		else { return }
+
+		sendFakeRequest()
+	}
+
+	/// Triggers one or more fake requests over a time interval of multiple seconds.
+	/// - Parameters:
+	///   - completion: called after all requests were triggered. Currently, only required when running in background mode to avoid terminating before the requests were made.
+	private func sendFakeRequest(_ completion: (() -> Void)? = nil) {
+		let service = exposureSubmissionService ?? ENAExposureSubmissionService(diagnosiskeyRetrieval: exposureManager, client: client, store: store)
+		let group = DispatchGroup()
+
+		for i in 0..<Int.random(in: minNumberOfSequentialPlaybooks...maxNumberOfSequentialPlaybooks) {
+			let delay = Int.random(in: minDelayBetweenSequentialPlaybooks...maxDelayBetweenSequentialPlaybooks)
+			group.enter()
+			DispatchQueue.global().asyncAfter(deadline: .now() + .seconds(i * delay)) {
+				service.fakeRequest()
+				group.leave()
```

Comment:
fakeRequest() always returns .fakeResponse (but only after the requests completed), at this point we do not really care about the result of the request but trigger it as fire and forget. I did not want to risk here that the background task runs into trouble because of slow network or whatever could fail during the fakeRequest call.

Reply:
What I concern is, the `service.fakeRequest()` may take some time. You ignore the completion of fakeRequest, so the code, `completion` will be called earlier when the fakeRequest not done yet.
```swift
 		group.notify(queue: .global()) {
			completion?()
		} 
```

Reply:
@johannesrohwer I got the reason from the call. Thanks.